(АКА-12) Сделал удаление локации из таблицы и её обновление после редактирования таргета

belannik opened the pull request:
+ Парочка изменений:

- Теперь удаление ячеек реализовано только правым свайпом

- Переопределять размеры у дефолтной табличной вьюшки удаления это больно, поэтому пришлось изменить подход к ячейкам. Теперь каждая ячейка - новая секция. Нужно для того, что бы определить расстояние между ними

Всё остальное должно быть понятно из коммитов :)

Commit:
(AKA-12) Убрал private у свойства даты создания

Commit:
(AKA-12) Добавил метод замены для SmartTargetCollection

Commit:
(AKA-12) Обновил метод возврата detailScene

Commit:
(AKA-12) Немного переверстал ячейку

Commit:
(АКА-12) Добавил методы для detailPresenter

Commit:
(AKA-12) Убрал лишний метод из роутера

Commit:
(АКА-12) Удалил ненужную картинку из ассетов

Commit:
(АКА-12) Класс кнопок для detail screen'a

Commit:
(AKA-12) удалил вьюшку, которая больше не нужна

Commit:
(АКА-12) DetailViewController

Добавил анимации и реализовал функционал

Commit:
(АКА-12) Изменил метод сохранения на удаление

Для SmartTargetList Scene

Commit:
(АКА-12) Обновил презентер для таблицы

Commit:
(АКА-12) Обновил роутер для таблицы

Commit:
(АКА-12) Обновил viewcontroller таблицы

Commit:
(АКА-12) Добавлял и удалял файлы

Commit:
Merge branch 'develop' of https://github.com/FocusStartSpb/Map into feature/AKA-12-editing-and-deleting-for-table-view-scene

FlightyNEO commented on Map/Business layer/Models/SmartTargetCollection.swift:
```diff
@@ -16,6 +16,7 @@ protocol ISmartTargetCollection: Codable
 	@discardableResult func put(_ smartTarget: SmartTarget) -> Int
 	func add(_ smartTargets: [SmartTarget])
 	@discardableResult func remove(atUID uid: String) -> Int?
+	func replace(on new: SmartTarget)
```
Функция put() заменит элемент, если элемент с таким же uid уже есть

belannik replied:
Уже сделано

FlightyNEO commented on Map/Business layer/SceneBuilder.swift:
```diff
@@ -92,12 +92,14 @@ final class SceneBuilder
 		let viewController = DetailTargetViewController(presenter: presenter,
 														router: router,
 														smartTargetEditable: smartTargetListViewController.isEditing)
+		presenter.attachViewController(detailTargetViewController: viewController)
```
Это через инициализатор презента можно передавить

FlightyNEO commented on Map/Scenes/Smart target list scene/Cell/SmartTargetTableViewCell.swift:
```diff
@@ -125,15 +129,16 @@ final class SmartTargetTableViewCell: UITableViewCell
+		guard let dateOfCreation = dateOfCreation else { return "" }
+		dateFormatted.dateStyle = .medium
+		let dateString = dateFormatted.string(from: dateOfCreation)
+		return dateString
```
Для этого я экстеншен создал, посмотри пожалуйста в файл Formatter+Extension.swift

FlightyNEO commented on Map/Scenes/Smart target list scene/Cell/SmartTargetTableViewCell.swift:
```diff
@@ -142,6 +147,6 @@ extension SmartTargetTableViewCell: ISmartTargetTableViewCell
 		self.titleLabel.text = smartTarget?.title
 		self.addressLabel.text = smartTarget?.address
-		self.timeOfCreationLabel.text = "12.12.2012"
+		self.timeOfCreationLabel.text = dateFormat(dateOfCreation: smartTarget?.dateOfCreated)
```
Выше написал про экстеншен к Forrmatter
Formatter.medium.string(from: smartTarget?.dateOfCreated ?? Date())

FlightyNEO commented on Map/Scenes/Smart target list scene/Detail target scene/DetailTargetPresenter.swift:
```diff
@@ -6,21 +6,34 @@
 {
 	private let smartTarget: SmartTarget
+	private let smartTargetCollection: ISmartTargetCollection
+	private weak var viewController: DetailTargetViewController?
```
Зачем weak?

FlightyNEO replied:
Тут weak не нужен и VC через инициализатор можно передавать

FlightyNEO commented on Map/Scenes/Smart target list scene/Detail target scene/DetailTargetPresenter.swift:
```diff
@@ -6,21 +6,34 @@
+		let formatter = DateFormatter()
+		formatter.dateFormat = "MMMM dd, yyyy 'at' hh:mm:ss"
+		let dateString = formatter.string(from: dateOfCreated)
+		return dateString
```
Опять же в экстеншен можно вынести

FlightyNEO commented on Map/Scenes/Smart target list scene/Detail target scene/DetailTargetViewController.swift:
```diff
@@ -26,15 +32,48 @@ final class DetailTargetViewController: UIViewController
 	private let router: IDetailTargetRouter
-	private let smartTargetEditable: Bool
+	private var smartTargetEditable = false {
+		didSet {
```
минор. Может декомпозировать этот код на 2 функции?

belannik replied:
Можно было бы и в экшенах у кнопок проставлять значения.
Я специально для этого вывел переменную, что бы всё делалось в одном месте. Мне показалось что так более наглядно

FlightyNEO commented on Map/Scenes/Smart target list scene/Detail target scene/DetailTargetViewController.swift:
```diff
@@ -198,6 +285,22 @@ final class DetailTargetViewController: UIViewController
+		//do something
+	}
+
+	private func mapViewNotEditable() {
```
NotEditable == Uneditable ?

Commit:
(АКА-12) Добавил экстеншен для форматтера

Commit:
(АКА-12) внёс правки в MapRouter & MapVC

Добавил navigationController в метод делегата у VC
Т.к. метод не проходил guard из-за того, что ранее SmartTargetListViewController обернулся в NavigationController
+ внёс аналогичные изменения в логику роутера

Commit:
(АКА-12) изменения в презентере детейла

Подменил форматтеры из экстеншена + немножко изменил метод saveChanges

Commit:
(АКА-12) Изменена логика DetailTargetRouter'a

Commit:
(АКА-12) Правки в VIP-цикл

Commit:
(АКА-12) сделал доступным сеттер для dataStore

Commit:
(АКА-12) добавил методы и свойства в VC таблицы

Для отлавливания отредактированного таргета и корректного его отображения

Commit:
Merge branch 'develop' of https://github.com/FocusStartSpb/Map into feature/AKA-12-editing-and-deleting-for-table-view-scene

Commit:
(АКА-12) Убрал лишние методы
